pkg/storage/stores/shipper/bloomshipper/fetcher.go

```diff
@@ -182,53 +248,70 @@ func (f *Fetcher) processTask(ctx context.Context, task downloadRequest[BlockRef
 		return
 	}
 
-	result, err := f.fetchBlock(ctx, task.item)
+	// check if block was fetched while task was waiting in queue
+	result, exists, err := f.getBlockDir(ctx, task.key)
 	if err != nil {
 		task.errors <- err
 		return
 	}
 
+	// return item from cache
+	if exists {
+		level.Debug(f.logger).Log("msg", "send download response", "reason", "cache")
+		task.results <- downloadResponse[BlockDirectory]{
+			item: result,
+			key:  task.key,
+			idx:  task.idx,
+		}
+		return
+	}
+
+	// fetch from storage
+	result, err = f.fetchBlock(ctx, task.item)
+	if err != nil {
+		task.errors <- err
+		return
+	}
+
+	// return item from storage
+	level.Debug(f.logger).Log("msg", "send download response", "reason", "storage")
 	task.results <- downloadResponse[BlockDirectory]{
 		item: result,
 		key:  task.key,
 		idx:  task.idx,
 	}
 }
 
-// fetchBlock resolves a block from three locations:
-//  1. from cache
-//  2. from file system
-//  3. from remote storage
-func (f *Fetcher) fetchBlock(ctx context.Context, ref BlockRef) (BlockDirectory, error) {
+// getBlockDir resolves a block directory without fetching the block from
+// remote object storage. Returns three arguments: the block directory, a
+// boolean whether the block was found in cache, and optionally an error.
+func (f *Fetcher) getBlockDir(ctx context.Context, key string) (BlockDirectory, bool, error) {
 	var zero BlockDirectory
 
 	if ctx.Err() != nil {
-		return zero, errors.Wrap(ctx.Err(), "fetch block")
+		return zero, false, errors.Wrap(ctx.Err(), "get block")
 	}
 
-	keys := []string{f.client.Block(ref).Addr()}
-
-	_, fromCache, _, err := f.blocksCache.Fetch(ctx, keys)
+	_, fromCache, _, err := f.blocksCache.Fetch(ctx, []string{key})
 	if err != nil {
-		return zero, err
+		return zero, false, err
 	}
 
 	// item found in cache
 	if len(fromCache) == 1 {
 		f.metrics.blocksFetchedSize.WithLabelValues(sourceCache).Observe(float64(fromCache[0].Size()))
-		return fromCache[0], nil
+		return fromCache[0], true, nil
 	}
 
-	fromLocalFS, _, err := f.loadBlocksFromFS(ctx, []BlockRef{ref})
```

Comment:
Do we no longer load blocks from the FS _at all_? Is this an oversight? It looks like we've removed the `fs` from the load path in all cases

I think querying bloom-gws should query against all local blocks (in mem or on disk)

Reply:
Since we load blocks into cache on startup ([cache.go](https://github.com/grafana/loki/blob/main/pkg/storage/stores/shipper/bloomshipper/cache.go#L55-L59)) we don't need to look at the filesystem at query/runtime.